src/OpaqueVal.cc

```diff
@@ -193,16 +205,189 @@ StringValPtr HashVal::DoGet() {
 
 HashVal::HashVal(OpaqueTypePtr t) : OpaqueVal(std::move(t)) { valid = false; }
 
-MD5Val::MD5Val() : HashVal(md5_type) { memset(&ctx, 0, sizeof(ctx)); }
+namespace {
+
+constexpr size_t MD5VAL_STATE_SIZE = sizeof(MD5_CTX);
+
+constexpr size_t SHA1VAL_STATE_SIZE = sizeof(SHA_CTX);
+
+constexpr size_t SHA256VAL_STATE_SIZE = sizeof(SHA256_CTX);
 
-MD5Val::~MD5Val() {
 #if ( OPENSSL_VERSION_NUMBER < 0x30000000L ) || defined(LIBRESSL_VERSION_NUMBER)
-    if ( IsValid() )
-        EVP_MD_CTX_free(ctx);
-#endif
+
+// -- MD5
+
+auto* to_native_ptr(MD5Val::StatePtr ptr) { return reinterpret_cast<EVP_MD_CTX*>(ptr); }
+
+auto* to_digest_ptr(MD5Val::StatePtr ptr) { return reinterpret_cast<detail::HashDigestState*>(ptr); }
+
+void do_init(MD5Val::StatePtr& ptr) { ptr = reinterpret_cast<MD5Val::StatePtr>(detail::hash_init(detail::Hash_MD5)); }
+
+void do_clone(MD5Val::StatePtr out, MD5Val::StatePtr in) { hash_copy(to_digest_ptr(out), to_digest_ptr(in)); }
+
+void do_feed(MD5Val::StatePtr ptr, const void* data, size_t size) {
+    detail::hash_update(to_digest_ptr(ptr), data, size);
+}
+
+void do_get(MD5Val::StatePtr ptr, u_char* digest) { detail::hash_final_no_free(to_digest_ptr(ptr), digest); }
+
+std::pair<const char*, size_t> do_get_bytes(MD5Val::StatePtr ptr) {
```

Comment:
Given that the callers of `do_get_bytes()` all subsequently create a std::string, could this just return a std::string directly (though internal, also makes ownership questions obvious)

Reply:
And nit, there's `do_get`, do wonder if this should be `do_serialize()` and `do_unserialize()` such that it aligns with where they are used and not too close to the existing ones.